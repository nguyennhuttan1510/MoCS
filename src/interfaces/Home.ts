export interface IListTable {
    handleSelectTable: Function,
    listTable: order[]
}

export interface food  {
    id: number,
    name: string,
    count: number,
    price: number,
    discount: number,
    total: number,
    chef: any,
    status: "Order"| "Pending" | "Done"
}

export interface order {
    isMakeFood?: boolean,
    id: number,
    name: string,
    menu: Array<food>,
    total_cost: number,
};

export interface ITableDetail {
    order?: order,
    handleAddMenu: Function,
    handleRemoveMenu: Function,
    onClose: Function

}

export interface IChefDetail {
    table?: order,
}