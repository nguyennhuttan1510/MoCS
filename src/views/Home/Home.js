import React from "react";
import { useDispatch, useSelector } from "react-redux";

import "../../style/_Home.scss";

import TableDetail from "views/Home/components/TableDetail";
import ListTable from "views/Home/components/ListTable";
import ChefDetail from "views/Home/components/ChefDetail";

import { table, setIsDetail } from "Reduces/dashboard";

import {
  handleAddTable,
  handleAddMenuTable,
  handleRemoveFood,
} from "util/socket/actionHome";
import { openNotificationWithIcon } from "components/Notification/Notification";

const Home = (props) => {
  const state = useSelector((state) => state.dashboard);
  const { isRedirectDetail } = state;
  const profile = useSelector((state) => state.staffs.profile);
  const tableCurrent = state.table;
  const tablesActive = state.data;
  const dispatch = useDispatch();

  const handleGetTableDetail = () => {
    if (!tablesActive) return;
    const result = tablesActive.filter((item) => item.id === tableCurrent.id);
    return result[0];
  };

  const onClose = (isClose) => {
    dispatch(setIsDetail(false));
    dispatch(table(false));
  };

  const handleSelectTable = (objTable) => {
    if (profile.position === "Staff") {
      // IF POSITION IS STAFF => TABLE IS CREATED
      if (objTable?.id) {
        const initObjTable = {
          id: objTable.id,
          name: objTable.name,
          menu: [],
          total_cost: 0,
        };
        handleAddTable(initObjTable);
      }
    }
    dispatch(setIsDetail(true));
    dispatch(table(objTable));
  };

  const handleRemoveMenu = (objFood, idTable) => {
    const payload = { id: idTable, food: objFood };
    handleRemoveFood(payload);
  };

  const handleAddMenu = (objFood, idTable) => {
    if (profile.position === "Admin") {
      openNotificationWithIcon(
        "error",
        "Error",
        "Admin isn't allow do this action"
      );
    }
    const total = (objFood?.price * (100 - objFood.discount)) / 100;
    const menu = {
      //a new food is added into list menu
      id: objFood?.id,
      name: objFood?.name,
      count: 1,
      price: objFood?.price,
      discount: objFood.discount,
      total: total,
      chef: false,
      status: "Order", // Order| Pending | Done
    };
    const payload = {
      id: idTable,
      food: menu,
    };
    handleAddMenuTable(payload);
  };

  return (
    <>
      {isRedirectDetail ? (
        profile.position === "Staff" || profile.position === "Admin" ? (
          <TableDetail
            order={handleGetTableDetail()}
            handleAddMenu={handleAddMenu}
            handleRemoveMenu={handleRemoveMenu}
            onClose={onClose}
          />
        ) : (
          <ChefDetail table={handleGetTableDetail()} />
        )
      ) : (
        <ListTable
          listTable={tablesActive}
          handleSelectTable={handleSelectTable}
        />
      )}
    </>
  );
};

export default Home;
