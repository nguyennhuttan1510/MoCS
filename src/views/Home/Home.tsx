import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { useDispatch, useSelector } from 'react-redux';

import '../../style/_Home.scss'

import TableDetail from 'views/Home/components/TableDetail';
import ListTable from 'views/Home/components/ListTable';
import ChefDetail from 'views/Home/components/ChefDetail';

import { food, order } from '../../interfaces/Home';

import { addMenuOfTable, addTable, table, setIsDetail, removeMenuOfTable } from 'Reduces/dashboard'

import { handleAddTable, handleAddMenuTable, handleRemoveFood } from 'util/socket/actionHome';
import axios from 'axios';
// import { useHistory } from 'react-router-dom';


const Home: React.FC = props => {
    // const history = useHistory();

    const state = useSelector((state: any) => state.dashboard);
    const { isRedirectDetail } = state;
    const profile = useSelector((state: any) => state.staffs.profile);
    const tableCurrent = state.table
    const tables: order[] = state.data


    // const [isDetail, setIsDetail] = useState<boolean>(false);
    const [orders, setOrders] = useState<Array<order>>(tables);
    const dispatch = useDispatch();

    useEffect(() => {
        setOrders(tables);
    }, [tables]);

    const handleGetTableDetail = () => {
        if (!orders) return
        const result = orders.filter((item) => item.id === tableCurrent.id);
        return result[0];
    }

    const onClose = (isClose: boolean) => {
        dispatch(setIsDetail(false))
        dispatch(table(false));
    }

    const handleSelectTable = (objTable: order) => {
        if (profile.position === "Staff") {               // IF POSITION IS STAFF => TABLE IS CREATED
            if (objTable?.id) {
                const initObjTable = {
                    id: objTable.id,
                    name: objTable.name,
                    menu: [],
                    total_cost: 0
                }
                const result = orders.filter((item) => item.id === objTable.id);
                console.log(result)
                console.log("into handleSelectTable");
                if (result.length === 0) {
                    handleAddTable(initObjTable)
                }

            }
        }
        dispatch(setIsDetail(true));
        dispatch(table(objTable));

    }

    const handleRemoveMenu = (objFood: any, idTable: number) => {
        console.log(idTable)
        const payload = { id: idTable, food: objFood }
        handleRemoveFood(payload)
    }

    const handleAddMenu = (objFood: any, idTable: number) => {
        // TODO: DUMMY DISCOUNT
        const total = (objFood?.price * (100 - 20) / 100) //THIS IS DUMMY DATA DISCOUNT
        const menu = {                                    //a new food is added into list menu 
            id: objFood?.id,
            name: objFood?.name,
            count: 1,
            price: objFood?.price,
            discount: 20,
            total: total,
            chef: false,
            status: "Order"                 // Order| Pending | Done 
        }
        const payload = {
            id: idTable, food: menu
        }
        handleAddMenuTable(payload)
    }

    return (
        <>
            {isRedirectDetail ? (profile.position === "Staff" || profile.position === "Admin" ? <TableDetail order={handleGetTableDetail()} handleAddMenu={handleAddMenu} handleRemoveMenu={handleRemoveMenu} onClose={onClose} /> : <ChefDetail table={handleGetTableDetail()} />) : <ListTable listTable={orders} handleSelectTable={handleSelectTable} />}
        </>
    );
};

Home.propTypes = {

};

export default Home;