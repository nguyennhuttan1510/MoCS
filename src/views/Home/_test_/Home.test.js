import React from 'react';
import Home from '../Home';

import { mount, shallow } from 'enzyme';
import { AppContext } from 'layouts/AppContext';
import toJson from 'enzyme-to-json';

describe('Home', () => {
  it('should be render', () => {
    const wrapper = shallow(
      <AppContext.Provider>
        <Home />
      </AppContext.Provider>
    );
    console.log(wrapper.debug());
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
