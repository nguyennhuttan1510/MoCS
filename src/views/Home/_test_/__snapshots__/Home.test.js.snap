// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should be render 1`] = `<Home />`;
